File: inferlet-js/src/async-utils.ts
// Async utilities for WASI pollable futures

/**
 * Minimal pollable interface (matches wasi:io/poll Pollable)
 */
interface Pollable {
  ready(): boolean;
  block(): void;
}

/**
 * Generic future interface for WASI async operations.
 * Matches WIT resources like receive-result, subscription, blob-result.
 */
export interface WasiFuture<T> {
  pollable(): Pollable;
  get(): T | undefined;
}

/**
 * Awaits a WASI future by blocking on its pollable.
 * @param future The WASI future to await
 * @param errorMessage Error message if result is undefined
 * @returns The resolved value
 * @throws Error if the future returns undefined
 */
export function awaitFuture<T>(future: WasiFuture<T>, errorMessage: string): T {
  const pollable = future.pollable();
  pollable.block();

  const result = future.get();
  if (result === undefined) {
    throw new Error(errorMessage);
  }
  return result;
}

File: sdk/python/src/inferlet_py/async_utils.py
"""
Async utilities for WASI pollable futures.
"""

from typing import Protocol, TypeVar, runtime_checkable

from wit_world.imports.poll import Pollable

T = TypeVar("T")


@runtime_checkable
class WasiFuture(Protocol[T]):
    """Generic interface for WASI async operations."""

    def pollable(self) -> Pollable: ...
    def get(self) -> T | None: ...


def await_future(future: WasiFuture[T], error_message: str) -> T:
    """
    Awaits a WASI future by blocking on its pollable.

    Args:
        future: The WASI future to await
        error_message: Error message if result is None

    Returns:
        The resolved value

    Raises:
        RuntimeError: If the future returns None
    """
    pollable = future.pollable()
    pollable.block()

    result = future.get()
    if result is None:
        raise RuntimeError(error_message)
    return result
